(GH-23) Initial docs for PCT (#25)
puppetlabs/pdkgo: docs/pct/README.md

* [Writing Your Own Templates](#writing-templates)
  * [Dos and Don'ts](#dos-and-don-ts)
  * [Structure](#structure)
    * [pct-config.yml](#pct-configyml)
  * [Templating Language](#templating-language)
  * [Testing](#testing)
  * [Sharing](#sharing)

## Overview

`Puppet Content Templates` (PCT) allow you to quickly create Puppet content from scaffolding templates built by Puppet and the Community.

> PCT is currently in an EXPERIMENTAL phase and feedback is encouraged via [pdk/discussions](https://github.com/puppetlabs/pdk/discussions) and starting a `feedback` post.

## Getting Started

Install the `exprimental` PCT release available from [github.com/puppetlabs/pdkgo/releases/](https://github.com/puppetlabs/pdkgo/releases/)

Templates currently come in 2 flavours: `project` and `item`.

* A `project` is a template containing many files in a particular structure. _These are great starting points._
* An `item` is a template that will supplement a project or existing content.

### pct new

PCT is available through the `pct new` command:

``` bash
pct new <template>
```

Replace `<template>` with the `name` of the template containing the content you want.

> By default the `new <template>` function will use the directory name of yor current working directory to "name" your new content.

Reviewer: ```suggestion
By default the `new <template>` function will use the directory name of your current working directory to "name" your new content.
```

```bash
           DISPLAYNAME          |          NAME           |  TYPE
--------------------------------+-------------------------+----------
                                |                         |
  Bolt Plan                     | bolt-plan               | item
  Bolt Project                  | bolt-project            | project
  Bolt PowerShell Task          | bolt-pwsh-task          | item
  Bolt YAML Plan                | bolt-yaml-plan          | item
  Puppet Module Managed Gemfile | git-attributes          | item
  Puppet Class                  | puppet-class            | item
  Puppet Content Template       | puppet-content-template | project
  Puppet Defined Type           | puppet-defined-type     | item
  Puppet Fact                   | puppet-fact             | item
  Puppet Module                 | puppet-module           | project
  Puppet Resource API Provider  | rsapi-provider          | item
  Puppet Resource API Transport | puppet-transport        | item
```

Using the available templates above, its time to generate some content.

``` bash
pct new <template>
```

Replace `<template>` with the `name` of the template containing the content you want.

> By default the `new <template>` function will use the directory name of yor current working directory to "name" your new content.

Reviewer: ```suggestion
By default the `new <template>` function will use the directory name of your current working directory to "name" your new content.
```

> :memo: Folders within the `content` directory can also use the `{{pct_name}}` variable

Example template file names:

``` bash
myConfig.json.tmpl
{{pct_name}}_spec.rb
```

> :memo: One, all or none of the files can be templated.

#### pct-config.yml

Format of pct-config.yml

``` yaml
---
template:
  id: <a unique name>
  type: <type>
  display: <a human readable name>
  version: <semver>
  url: <url to project>

<template parameters>
```

Example pct-config.yml:

``` yaml
---
template:
  id: example-template
  type: project
  display: Example
  version: 0.1.0
  url: https://github.com/puppetlabs/pct-example
```

Example structure for `example-template`:

``` bash
> tree ~/.pdk/pct/example-template-params
/Users/me/.pdk/pct/example-template-params
├── content
│   └── example.txt.tmpl
└── pct-config.yml
```

Example pct-config.yml with parameters:

``` yaml
---
template:
  id: example-template-params
  type: project
  display: Example with Parameters
  version: 0.1.0
  url: https://github.com/puppetlabs/pct-example-params

example_template:
  foo: "bar"
  isPuppet: true
  colours:
  - "Red"
  - "Blue"
  - "Green"

```

> ### Templating Language

Reviewer: Just a minor readability thing - given the above `yaml` snippet is related to the `go` snippet below, the **Templating Language** breaks up the flow a bit.

Author: I'm not sure there is any easy way to fix that other than copy/paste that code again.

Reviewer 2: I had exactly the same suggestion as Ciaran. I hear your point about it kinda belonging in both places but the parameters topic is more related to the templating language section than it is for the file structure section. I vote for moving it entirely and not including it where it is at the moment

# Puppet Content Templates

* [Overview](#overview)
* [Getting Started](#getting-started)
  * [pct new](#pct-new)
  * [Template Updates](#template-updates)
  * [More Templates](#more-templates)
* [Writing Your Own Templates](#writing-templates)
  * [Dos and Don'ts](#dos-and-donts)
  * [Structure](#structure)
    * [pct-config.yml](#pct-configyml)
  * [Templating Language](#templating-language)
  * [Testing](#testing)
  * [Sharing](#sharing)
* [Overriding Template Defaults](#overriding-template-defaults)

## Overview

> `Puppet Content Templates` (PCT) allow you to quickly create Puppet content from scaffolding templates built by Puppet and the Community.

Reviewer 2: You shouldn't put a code block around the 'Puppet Content Templates' phrase unless you're going to do it for the entire document. In general we use code blocks for language keywords not the name of the language itself.

Reviewer 2: I think we can use some of the language in the specification document to flesh out the intro here. We have the opportunity to describe the purpose as well as show why to use the product in the intro

> > :warning: PCT is currently in an EXPERIMENTAL phase and feedback is encouraged via [pdk/discussions](https://github.com/puppetlabs/pdk/discussions) and starting a `feedback` post.

Reviewer 2: Should we make a discussions tab in the pdkgo repo instead of mixing in the pdk repo?

Author: Sounds good

## Getting Started

> Install the `exprimental` PCT release available from [github.com/puppetlabs/pdkgo/releases/](https://github.com/puppetlabs/pdkgo/releases/)

Reviewer 2: Would use the word 'download' here as we don't yet have an installer :)

Reviewer 3: this may be on a do to list or covered elsewhere, but how do we set up the experimental PCT?

Author: > this may be on a do to list or covered elsewhere, but how do we set up the experimental PCT?

We don't have the PCT in this state. I've written some of this documentation based on where i think we need to be in order to release an LA and to stimulate discussion.

Templates currently come in 2 flavours: `project` and `item`.

> * A `project` is a template containing many files in a particular structure. _These are great starting points._

Reviewer 2: ```suggestion
* A `project` is a template containing many files in a particular structure. They create a ready-to-run structures to start using a Puppet product. _These are great starting points._ You can create a boilerplate empty starter Puppet Module or a fully customized Puppet Module with specialized CI files and RSAPI providers.
```

> * An `item` is a template that will supplement a project or existing content.

Reviewer 2: ```suggestion
* An `item` is a template that will supplement a project or existing content. These could be language features like a Puppet class or single files like a Git ignore file.
```

> ### pct new

Reviewer 2: ```suggestion
### Creating content using Puppet Content Templates
```

I would know to look for creating if I didn't know anything about the tool, where as I wouldn't know to look for `pct new`

> PCT is available through the `pct new` command.

Reviewer 2: ```suggestion
You can deploy a PCT through the `pct new` command.
```

I read this as 'Puppet Content Template is available...' and the singular/plural mismatch was jarring.

> The `--list` or `-l` flag displays a list of locally available templates located in the default template location `$HOME/.pdk/pct/`. The list of templates is also available by calling `pct new` without flags.

Reviewer 2: ```suggestion
To display at list of available templates currently installed on your system, use the `--list` or `-l` flag. This reads from the default template location `$HOME/.pdk/pct/`. The list of installed templates is also available by calling `pct new` without flags.
```

``` bash
pct new
pct new --list
```

Example output:

```bash
           DISPLAYNAME          |          NAME           |  TYPE
--------------------------------+-------------------------+----------
                                |                         |
  Bolt Plan                     | bolt-plan               | item
  Bolt Project                  | bolt-project            | project
  Bolt PowerShell Task          | bolt-pwsh-task          | item
  Bolt YAML Plan                | bolt-yaml-plan          | item
  Puppet Module Managed Gemfile | git-attributes          | item
  Puppet Class                  | puppet-class            | item
  Puppet Content Template       | puppet-content-template | project
  Puppet Defined Type           | puppet-defined-type     | item
  Puppet Fact                   | puppet-fact             | item
  Puppet Module                 | puppet-module           | project
  Puppet Resource API Provider  | rsapi-provider          | item
  Puppet Resource API Transport | puppet-transport        | item
```

Using the available templates above, its time to generate some content.

``` bash
pct new <template>
```

> 

Reviewer 2: ```suggestion

> 📝 You can use shell tab completion to resolve the template names without having to remember the exact spelling by enabling shell completions for your current shell. See [shell completions](#shell-completions) section for more information.
```

Reviewer 2: Still think this would be a good addition

Replace `<template>` with the `name` of the template containing the content you want.

> By default the `new <template>` function will use the directory name of your current working directory to "name" your new content.

Reviewer 2: ```suggestion
By default the `pct new <template>` command will use the directory name of your current working directory to "name" your new content.
```

To override this behavour use the `--name` or `-n` flag.

``` bash
pct new <template> --name MyProject
```

> By default the `new <template>` function will output the template content to the current working directory.

Reviewer 2: ```suggestion
By default the `pct new <template>` command will output the template content to the current working directory.
```

To override this behavour use the `--output` or `-o` flag.

``` bash
pct new <template> --output /path/to/your/project
```

> > :memo: Not all templates require a `name` and providing one will have no effect on the generated content.

Reviewer 2: ```suggestion
> :memo: Not all templates require a `name`. If a template doesn't require on, providing a value to the `--name` parameter will have no effect on the generated content.
```

> Example usage:

Reviewer 2: ```suggestion
### Example workflow
```

I think this deserves it's own sub section with some verbage about how the building block approach can stitch together a set of content into a final product

``` bash
> cd /home/me/projects/MyBoltProject
> pct new bolt-project

> 

Reviewer 2: These are two separate examples, but contained in the same code block. That was confusing to me, I suggest splitting the two and giving a one sentence 'This workflow creates a `<foo>` that has `<bar>` and `<baz>` components' kinda thing

> pct new puppet-module -n MyNewProject -o /home/me/projects/
> cd /home/me/projects/MyNewProject
> pct new puppet-fact -n ApplicationVersion
> pct new rsapi-provider -n Awesomething
> pct new puppet-transport -n AwesomethingApi
```

### Template Updates

> `pct new` will **NOT** update your code to a newer version of a template.

Reviewer 2: ```suggestion
At this time `pct new` will **NOT** update an existing code to a newer version of a template.
```

If you run a `pct new` command using a `project` template, the project will replace the content within the output directory with the template code.

If you run a `pct new` command using an `item` template, the item will suppliment the content within the output directory with the template code. If files / folders that are named the same as the template content already exist, it will overwite this content.

## Writing Templates

> ### Dos and Don'ts

Reviewer 2: This is a great start, and something we'll add to over time. I think this section should be after the structure and example sections in this block

example_template:
  foo: "bar"
  isPuppet: true
  colours:
  - "Red"
  - "Blue"
  - "Green"

```

### Templating Language

PCT uses [Go's templating language](https://golang.org/pkg/text/template/#hdr-Actions).

In the above template `example-template-params` the parameters can be accessed in a `.tmpl` file like so:

``` go
{{.example_template.foo}}
{{.example_template.isPuppet}}
{{.example_template.colours}}
```

outputs:

``` text
bar
<no value>
[Red Blue Green]
```

As a template author you can chose your own parameters and parameter structure so long as it is [valid YAML](https://yaml.org/spec/1.2/spec.html). Then utilise the GO templating language to display or iterate over these.

For most templates, we believe that you can do most of the things you would want with these common template controls:

``` go
// Outputs the value of `foo` defined within pct.yml
{{.example_template.foo}}

// A conditional
{{if .example_template.isPuppet}}
 "boo :("
{{else}}
 "yay!"
{{end}}

// Loops over all "colours" and renders each using {{.}}
{{range .example_template.colours}} {{.}} {{end}}
```

For more examples look at the existing templates provided at `$HOME/.pdk/pct/`

> ### Testing

Reviewer 2: I'd rather not include the section than leave it empty. I thought the markdown rendered incorrectly instead of it being empty 😂

Author: This was more of a holder to ask the question of how to best test a template.

### Sharing

:shipit:

## Overriding Template Defaults

Perhaps you use a template often and find that you set the same values over and over?
As a template user, you can choose to override the default values specified by a template author.

To view the default parameters for a template run:

``` bash

> pct new <template> --show-defaults

Reviewer 2: We already have `--list` which when given a template id displays a short amount of info (at the moment). I propose instead of a new parameter we use `--list` and extend what it shows when a single template is returned

```suggestion
pct new --list <template>
```

We can table this discussion by leaving this section out and ticketing the work to create the feature and document at the same time, instead of writing the document with the aspiration of making the feature

  * [Templating Language](#templating-language)
  * [Testing](#testing)
  * [Sharing](#sharing)
* [Overriding Template Defaults](#overriding-template-defaults)

## Overview

Puppet Content Templates (PCT) codify a structure to produce content for any Puppet Product that can be authored by Puppet Product Teams or external users without direct help of the PDK team.

PCT can create any type of a Puppet Product project: Puppet control repo, Puppet Module, Bolt project, etc. It can create one or more independent files, such as CI files or gitignores. This can be as simple as a name for a Puppet Class, a set of CI files to add to a Puppet Module, or as complex as a complete Puppet Control repo with roles and profiles.

These are meant to be ready-to-run, which means they put everything needed for a user to run the project from the moment after creation. This solves the 'blank page' problem, where a few files are in place but the user does not know what the next steps are.

> :warning: PCT is currently in an EXPERIMENTAL phase and feedback is encouraged via [pdkgo/discussions](https://github.com/puppetlabs/pdkgo/discussions) and starting a `feedback` post.

## Getting Started

> Install the `exprimental` PCT release available from [github.com/puppetlabs/pdkgo/releases/](https://github.com/puppetlabs/pdkgo/releases/)

Reviewer: ```suggestion
Install the `experimental` PCT release available from [github.com/puppetlabs/pdkgo/releases/](https://github.com/puppetlabs/pdkgo/releases/)
```

Templates currently come in 2 flavours: `project` and `item`.

* A `project` is a template containing many files in a particular structure. They create a ready-to-run structures to start using a Puppet product. _These are great starting points._ You can create a boilerplate empty starter Puppet Module or a fully customized Puppet Module with specialized CI files and RSAPI providers.
* An `item` is a template that will supplement a project or existing content. These could be language features like a Puppet class or single files like a Git ignore file.

### pct new

PCT is available through the `pct new` command.

> The `--list` or `-l` flag displays a list of locally available templates located in the default template location `$HOME/.pdk/pct/`. The list of templates is also available by calling `pct new` without flags.

Reviewer: **Placeholder Note:** Based on the decision made [here](https://github.com/puppetlabs/pdkgo/issues/20#issuecomment-841178719), we should update the docs, accordingly

Reviewer 2: In progress

                                |                         |
  Bolt Plan                     | bolt-plan               | item
  Bolt Project                  | bolt-project            | project
  Bolt PowerShell Task          | bolt-pwsh-task          | item
  Bolt YAML Plan                | bolt-yaml-plan          | item
  Puppet Module Managed Gemfile | git-attributes          | item
  Puppet Class                  | puppet-class            | item
  Puppet Content Template       | puppet-content-template | project
  Puppet Defined Type           | puppet-defined-type     | item
  Puppet Fact                   | puppet-fact             | item
  Puppet Module                 | puppet-module           | project
  Puppet Resource API Provider  | rsapi-provider          | item
  Puppet Resource API Transport | puppet-transport        | item
```

Using the available templates above, its time to generate some content.

``` bash
pct new <template>
```

Replace `<template>` with the `name` of the template containing the content you want.

By default the `new <template>` function will use the directory name of your current working directory to "name" your new content.

> To override this behavour use the `--name` or `-n` flag.

Reviewer: ```suggestion
To override this behaviour use the `--name` or `-n` flag.
```
